File: code/CountDistinct.java
import java.util.Arrays;
import java.util.Random;
import java.util.Scanner;

public class CountDistinct {
  public static void main(String[] args) {
    int n = 10; // number of estimates
    int b = 2; // size of bunch
    if (args.length >= 1) n = Integer.parseInt(args[0]); assert (n>=1);
    if (args.length >= 2) b = Integer.parseInt(args[1]); assert (b>=1);
    int a[] = new int[n];
    int r[] = new int[n];
    Random rand = new Random(123);
    for (int i = 0; i < n; ++i) a[i] = rand.nextInt();
    Scanner scan = new Scanner(System.in);
    while (scan.hasNext()) {
      int x = (int)scan.nextLong();
      for (int i = 0; i < n; ++i) r[i] = Math.max(r[i], rho(x+a[i]));
    }
    int m = (n + b - 1) / b;
    double avg[] = new double[m];
    for (int i = 0; i < n; ++i) avg[i/b] += Math.pow(2, r[i]);
    for (int j = 0; j < m; ++j) avg[j] /= b;
    Arrays.sort(avg);
    double e = (m % 2 == 1)? avg[m/2] : (avg[m/2]+avg[m/2-1])/2.0;
    System.out.printf("%.0f\n", e);
  }

  static int rho(int x) {
    if (x == 0) return 32;
    int r = 0;
    while (x % 2 == 0) { ++r; x >>= 1; }
    return r;
  }
}

File: code/CountDistinct.hs
-- ghc --make -O2 CountDistinct.hs -o CountDistinct-Haskell
-- TODO: fix -- this is pretty ugly
-- TODO: point out Data.HyperLogLog
module Main where

import Data.List
import Data.List.Split
import System.Random as R

main :: IO ()
main = interact (shownl . solve . map read . lines)

shownl :: Double -> String
shownl x = show x ++ "\n"

solve :: [Integer] -> Double
solve xs = combine (foldl oneIP zerorho xs)

zerorho :: [(Integer, Integer)]
zerorho = start ([], R.mkStdGen 123) 100

start (xs, _) 0 = xs
start (xs, g) n = start ((x, 0) : xs, ng) (n - 1)
  where (x, ng) = R.random g :: (Integer, R.StdGen)

oneIP :: [(Integer, Integer)] -> Integer -> [(Integer, Integer)]
oneIP ars x = map (oneHASHoneIP x) ars

oneHASHoneIP :: Integer -> (Integer, Integer) -> (Integer, Integer)
oneHASHoneIP x (a, r) = (a, max r (rho (a + x)))

combine :: [(Integer, Integer)] -> Double
combine = median . map avg . chunksOf 3 . map (2^) . map snd

avg :: [Integer] -> Double
avg xs = fromIntegral (foldl (+) 0 xs) / fromIntegral (length xs)
median xs = sort xs !! (length xs `div` 2)

rho :: Integer -> Integer
rho 0 = 32
rho x
  | x `mod` 2 == 1 = 0
  | otherwise = 1 + rho (x `div` 2)
